TIP-20 Transaction Payload with TIP-18 Output Types (#40)
iotaledger/tips: tips/TIP-0020/tip-0020.md

<table>
  <tr>
    <th>Name</th>
    <th>Type</th>
    <th>Description</th>
  </tr>
  <tr>
    <td>Payload Type</td>
    <td>uint32</td>
    <td>
      Set to <strong>value 6</strong> to denote a <i>TIP-20 Transaction Payload</i>.
    </td>
  </tr>
  <tr>
    <td valign="top">Essence <code>oneOf</code></td>
    <td colspan="2">
      <details open="true">
        <summary>Transaction Essence</summary>
        <blockquote>
          Describes the essence data making up a transaction by defining its inputs, outputs and an optional payload.
        </blockquote>
        <table>
          <tr>
            <td><b>Name</b></td>
            <td><b>Type</b></td>
            <td><b>Description</b></td>
          </tr>
          <tr>
            <td>Transaction Type</td>
            <td>uint8</td>
            <td>
              Set to <strong>value 1</strong> to denote a <i>TIP-20 Transaction Essence</i>.
            </td>
          </tr>
          <tr>
            <td>Network ID</td>
            <td>uint64</td>
            <td>

>               The unique value denoting whether the block was meant for mainnet, shimmer, testnet, or a private networks.

Reviewer: ```suggestion
              The unique value denoting whether the block was meant for mainnet, shimmer, testnet, or a private network. It consists of the first 8 bytes of the BLAKE2b-256 hash of the network name.
```

      </details>
    </td>
  </tr>
</table>

### Transaction Essence

The <i>Transaction Essence</i> of a <i>Transaction Payload</i> carries the inputs, outputs, and an optional payload. The <i>Transaction Essence</i> is an explicit type and therefore starts with its own <i>Transaction Essence Type</i> byte which is of value 1 for <i>TIP-20 Transaction Essence</i>.

#### Inputs

The `Inputs` field holds the inputs to consume in order to fund the outputs of the <i>Transaction Payload</i>. Currently, there is only one type of input, the <i>UTXO Input</i>. In the future, more types of inputs may be specified as part of protocol upgrades.

Each input must be accompanied by a corresponding <i>Unlock</i> at the same index in the <i>Unlocks</i> part of the <i>Transaction Payload</i>.

##### UTXO Input

A <i>UTXO Input</i> is an input which references an unspent output of a previous transaction. This UTXO is uniquely identified by its _Output ID_, defined by the _Transaction ID_ of the creating transaction together with corresponding output index. Each <i>UTXO Input</i> must be accompanied by an <i>Unlock</i> that is allowed to unlock the referenced output.

#### Inputs Commitment

> The _Inputs Commitment_ field of the _Transaction Essence_ is a cryptographic commitment to the content of the consumed outputs (inputs). The transaction itself only references inputs by _Output ID_, therefore the client has to be aware of the content to produce a semantically valid transaction. The commitment is a security mechanism that protects clients against [eclipse attacks that would result in loss of funds.](https://github.com/iotaledger/tips/discussions/51)

Reviewer 2: ```suggestion
The `Inputs Commitment` field of the _Transaction Essence_ is a cryptographic commitment to the content of the consumed outputs (inputs). It consists of the BLAKE2b-256 hash of the concatenated output hashes.

In the `Inputs` field, they are only referenced by  _Output ID_. While the _Output ID_ technically depends on the content of the actual output, a client has no way of validating this without access to the original transaction. For the `Inputs Commitment`, the client has to be aware of the outputs’ content in order to produce a semantically valid transaction. This protects clients against [eclipse attacks that would result in loss of funds](https://github.com/iotaledger/tips/discussions/51).
```
The formal definition is given under validation

#### Outputs

The `Outputs` field holds the outputs that are created by the <i>Transaction Payload</i>. There are different output types, but they must all have an `Amount` field denoting the number of IOTA coins to deposit.

> The following table lists all the output types that are currently supported as well as links to the corresponding specification. The _SigLockedSingleOutput_ as well as the _SigLockedDustAllowanceOutput_ introduced in Chrysalis Phase 2 [TIP-7](https://github.com/iotaledger/tips/blob/main/tips/TIP-0007/tip-0007.md) have been removed and are no longer supported.

Reviewer 2: ```suggestion
The following table lists all the output types that are currently supported as well as links to the corresponding specification. The _SigLockedSingleOutput_ as well as the _SigLockedDustAllowanceOutput_ introduced in Chrysalis Phase 2 [TIP-7](../TIP-0007/tip-0007.md) have been removed and are no longer supported.
```

| Output Name | Type Value | TIP                                                                                                          |
| ----------- | ---------- | ------------------------------------------------------------------------------------------------------------ |

> | Basic    | 3          | [draft TIP-18](https://github.com/lzpap/protocol-rfcs/blob/master/tips/TIP-0018/tip-0018.md#basic-output) |

Reviewer 2: ```suggestion
| Basic       | 3          | [draft TIP-18](https://github.com/lzpap/protocol-rfcs/blob/master/tips/TIP-0018/tip-0018.md#basic-output)   |
```
Allignment

| Alias       | 4          | [draft TIP-18](https://github.com/lzpap/protocol-rfcs/blob/master/tips/TIP-0018/tip-0018.md#alias-output)    |
| Foundry     | 5          | [draft TIP-18](https://github.com/lzpap/protocol-rfcs/blob/master/tips/TIP-0018/tip-0018.md#foundry-output)  |

> | NFT         | 6          | [draft TIP-18](https://github.com/lzpap/protocol-rfcs/blob/master/tips/TIP-0018/tip-0018.md#nft-output)      |

Reviewer 2: What do we do here? Leave TIP-18 as a draft pointing to the PR and then change it to the proper link, once that gets merged?

Author: Yes, my plan is to merge all the TIPs and then create PR that updates the refs present in the repo.

#### Payload

> The  _Transaction Essence_ itself can contain another payload as described in general in  [draft TIP-24](https://github.com/iotaledger/tips/pull/55). The [semantic validity](#semantic-validation) of the encapsulating _Transaction Payload_ does not have any impact on the payload.

Reviewer 2: ```suggestion
The  _Transaction Essence_ itself can contain another payload as described in general in  [TIP-24](../TIP-0024/tip-004.md). The [semantic validity](#semantic-validation) of the encapsulating _Transaction Payload_ does not have any impact on the payload.
```

## Validation

A <i>Transaction Payload</i> has different validation stages, since some validation steps can only be executed when certain information has (or has not) been received. We therefore distinguish between syntactic and semantic validation.

The different output types and optional output features introduced by [draft TIP-18](https://github.com/iotaledger/tips/pull/38) add additional constraints to the transaction validation rules, but since these are specific to the given outputs and features, they are discussed for each [output type](https://github.com/lzpap/protocol-rfcs/blob/master/tips/TIP-0018/tip-0018.md#output-design) and [feature type](https://github.com/lzpap/protocol-rfcs/blob/master/tips/TIP-0018/tip-0018.md#output-features) separately.

### Syntactic validation

Syntactic validation is checked as soon as the transaction has been received. It validates the structure but not the signatures of the transaction. If the transaction does not pass this stage, it must not be broadcast further and can be discarded right away.

The following criteria defines whether a payload passes the syntactical validation:

* Essence:

>   * `Transaction Type` value must denote a _Transaction Essence_.

Reviewer 2: ```suggestion
  * `Transaction Type` value must denote a _TIP-20 Transaction Essence_.
```

### Semantic validation

The Semantic validation of a _Transaction Payload_ is performed when its encapsulating block is confirmed by a milestone. The semantic validity of transactions depends on the order in which they are processed. Thus, it is necessary that all the nodes in the network perform the checks in the same order, no matter the order in which the transactions are received. This is assured by using the White-Flag ordering as described in  [TIP-2](../TIP-0002/tip-0002.md#deterministically-ordering-the-tangle).

Processing transactions according to the White-Flag ordering enables users to spend UTXOs which are created in the same milestone confirmation cone, as long as the spending transaction comes after the funding transaction in the aforementioned White-Flag order. In this case, it is recommended that users include the _Block ID_ of the funding transaction as a parent of the block containing the spending transaction.

The following criteria defines whether a payload passes the semantic validation:
* Each input must reference a valid UTXO, i.e. the output referenced by the input's `Transaction ID` and `Transaction Output Index` is known (booked) and unspent.

> * `Inputs Commitment` must be equal to the BLAKE2b-256 hash of the BLAKE2b-256 hashes of the serialized outputs that are referenced as inputs. The order of the serialized outputs must be the same as their order in `Inputs`.

Reviewer 2: ```suggestion
* `Inputs Commitment` must equal BLAKE2( BLAKE2(O<sub>1</sub>) || … || BLAKE2(O<sub>n</sub>) ), where O<sub>1</sub>, ..., O<sub>n</sub> are the complete serialized outputs referenced by the `Inputs` field in that order.
```

# Drawbacks

* The new transaction format is the core data type within the IOTA ecosystem. Changing it means that all projects need to accommodate it, including wallets, web services, client libraries and applications using IOTA in general. It is not possible to keep these changes backwards compatible, meaning that all nodes must upgrade to further participate in the network.
* It is not possible to produce a valid transaction without having access to the content of the consumed outputs.

# Rationale and alternatives

* _Inputs Commitment_ and _Network ID_ are both explicit fields of the transaction, while they could be made configuration
  parameters for the signature generating process. In this scenario the signature would be invalid if the parameters on client
  and network side mismatch. While this would reduce the size of a transaction, it would make it impossible to debug the
  reason for having an invalid signature and transaction. With the current solution we intend to optimize for ease of development.

> * Uniqueness of all inputs is kept as it prevents introducing double spends in the same transaction.

Reviewer 2: ```suggestion
```
I think that sentence should just be removed, right?

Author: In syntactic validation we have this rule:
> Each pair of Transaction ID and Transaction Output Index must be unique in the list of inputs.

Therefore, this sentence is still valid, no?

Reviewer 2: OK, fair enough

<table>
  <tr>
    <th>Name</th>
    <th>Type</th>
    <th>Description</th>
  </tr>
  <tr>
    <td>Payload Type</td>
    <td>uint32</td>
    <td>
      Set to <strong>value 0</strong> to denote a <i>Transaction Payload</i>.
    </td>
  </tr>
  <tr>
    <td valign="top">Essence <code>oneOf</code></td>
    <td colspan="2">
      <details open="true">
        <summary>Transaction Essence</summary>
        <blockquote>
          Describes the essence data making up a transaction by defining its inputs, outputs and an optional payload.
        </blockquote>
        <table>
          <tr>
            <td><b>Name</b></td>
            <td><b>Type</b></td>
            <td><b>Description</b></td>
          </tr>
          <tr>
            <td>Transaction Type</td>
            <td>uint8</td>
            <td>
              Set to <strong>value 0</strong> to denote a <i>Transaction Essence</i>.
            </td>
          </tr>
          <tr>
            <td>Network ID</td>
            <td>uint64</td>
            <td>

>               The unique value denoting whether the message was meant for mainnet, testnet, or a private networks. The <code>Network ID</code> should usually match the corresponding field of the encapsulating message, but this is not enforced.

Reviewer 3: There is no network ID in the message anymore.

---
tip: 20
title: Transaction Payload with TIP-18 Output Types
description: Add output types, unlocks, and output features from TIP-18 into Transaction Payload
author: Levente Pap (@lzpap) <levente.pap@iota.org>
discussions-to: https://github.com/iotaledger/tips/pull/40
status: Draft
type: Standards
layer: Core
created: 2021-11-18
requires: TIP-7 and TIP-18
---

# Summary

This TIP proposes a UTXO-based transaction structure consisting of all the inputs and outputs of a transfer. Specifically, this TIP defines a transaction payload for _blocks_ described in [TIP-24](../TIP-0024/tip-0024.md).

# Motivation

> There are several options on how transfers of IOTA coins can be embedded into the actual vertices of the Tangle. In the legacy IOTA protocol, each vertex corresponds to a single transaction, i.e. an input or an output. In order to support atomic transfers, input/output transactions have to be grouped into a so-called bundle and then either the entire bundle is applied to the ledger or none of its transactions.

Reviewer 3: ```suggestion
There are several options on how transfers of IOTA coins can be embedded into the actual vertices of the Tangle. In the legacy IOTA protocol, each vertex corresponds to a single transaction, i.e. an input or an output. In order to support atomic transfers, input/output transactions had to be grouped into a so-called bundle and then either the entire bundle is applied to the ledger or none of its transactions.
```

Author: The rest of the paragraph is in present tense

Reviewer 3: Yeah feel free to ignore, it just feels a bit weird speaking about legacy IOTA in present tense, can be a bit misleading idk

> The bundle concept has proven to be rather challenging in practices as validation can get very complex, especially in the context of reattachments. Therefore, this TIP proposes an alternative approach: It defines a self-contained transaction payload, containing the entire transfer, that is then embedded into a single Tangle block/vertex.

Reviewer 3: ```suggestion
The bundle concept has proven to be rather challenging in practice as validation can get very complex, especially in the context of reattachments. Therefore, this TIP proposes an alternative approach: It defines a self-contained transaction payload, containing the entire transfer, that is then embedded into a single Tangle block/vertex.
```

Using a UTXO-based model provides several benefits:
* Parallel validation of transactions.
* Easier double-spend detection, since conflicting transactions would reference the same UTXO.
* Replay-protection which is important when having reusable addresses. Replaying the same transaction would manifest itself as already being applied or existent and thus not have any impact.
* Balances are no longer strictly associated to addresses. This allows a higher level of abstraction and thus enables other types of outputs with particular unlock criteria.

Within a transaction using UTXOs, inputs and outputs make up the to-be-signed data of the transaction. The section unlocking the inputs is called the *unlock*. An unlock may contain a signature proving ownership of a given input's address and/or other unlock criteria.

The following image depicts the flow of funds using UTXO:

> ![UTXO flow](utxo.png)

Reviewer 3: (if in dark mode)

## Structure

### Serialized layout

A _Transaction Payload_ is made up of two parts:

> 1. The _Transaction Essence_ part which contains the inputs, outputs and an optional embedded payload.

Reviewer 3: Just a note: this is not really exhaustive

                <blockquote>
                  Describes an input which references an unspent transaction output to consume.
                </blockquote>
                <table>
                  <tr>
                    <td><b>Name</b></td>
                    <td><b>Type</b></td>
                    <td><b>Description</b></td>
                  </tr>
                  <tr>
                    <td>Input Type</td>
                    <td>uint8</td>
                    <td>
                      Set to <strong>value 0</strong> to denote an <i>UTXO Input</i>.
                    </td>
                  </tr>
                  <tr>
                    <td>Transaction ID</td>
                    <td>ByteArray[32]</td>
                    <td>The BLAKE2b-256 hash of the transaction payload containing the referenced output.</td>
                  </tr>
                  <tr>
                    <td>Transaction Output Index</td>
                    <td>uint16</td>
                    <td>The output index of the referenced output.</td>
                  </tr>
                </table>
              </details>
            </td>
          </tr>
          <tr>

>             <td>Inputs Commitment</td>

Reviewer 3: The rationale behind inputs commitment is never really explained in this TIP, maybe a small paragraph about it in the next section would be nice?

Author: Added an explanation

            <td>ByteArray[32]</td>
            <td>

>               BLAKE2b-256 hash of the serialized outputs referenced in <i>Inputs</i> by their <i>Output IDs</i> (<i>Transaction ID</i> || <i>Transaction Output Index</i>).

Reviewer 4: ```suggestion
              BLAKE2b-256 hash of the BLAKE2b-256 hashes of the serialized outputs referenced in <i>Inputs</i> by their <i>Output IDs</i> (<i>Transaction ID</i> || <i>Transaction Output Index</i>).
```

Reviewer 4: I would like to merklelize the input commitment, so we would have the chance to compress data in the ledger on nodes.

With this change a node owner could decide to remove the content of the metadata fields of the output in his own node database. If the node hashes the output before removing the data, and stores the hash, the check for correct input commitment would still pass.

            </td>
          </tr>
          <tr>
            <td>Outputs Count</td>
            <td>uint16</td>
            <td>The number of output entries.</td>
          </tr>
          <tr>
            <td valign="top">Outputs <code>anyOf</code></td>
            <td colspan="2">
              <details>

>                 <summary>Extended Output</summary>

Reviewer 3: ```suggestion
                <summary>Basic Output</summary>
```

#### Outputs

The `Outputs` field holds the outputs that are created by the <i>Transaction Payload</i>. There are different output types, but they must all have an `Amount` field denoting the number of IOTA coins to deposit.

The following table lists all the output types that are currently supported as well as links to the corresponding specification. The _SigLockedSingleOutput_ as well as the _SigLockedDustAllowanceOutput_ introduced in Chrysalis Phase 2 [TIP-7](https://github.com/iotaledger/tips/blob/main/tips/TIP-0007/tip-0007.md) have been removed and are no longer supported.

| Output Name | Type Value | TIP                                                                                                          |
| ----------- | ---------- | ------------------------------------------------------------------------------------------------------------ |

> | Extended    | 3          | [draft TIP-18](https://github.com/lzpap/protocol-rfcs/blob/master/tips/TIP-0018/tip-0018.md#extended-output) |

Reviewer 3: ```suggestion
| Basic    | 3          | [draft TIP-18](https://github.com/lzpap/protocol-rfcs/blob/master/tips/TIP-0018/tip-0018.md#basic-output) |
```

| Foundry     | 5          | [draft TIP-18](https://github.com/lzpap/protocol-rfcs/blob/master/tips/TIP-0018/tip-0018.md#foundry-output)  |
| NFT         | 6          | [draft TIP-18](https://github.com/lzpap/protocol-rfcs/blob/master/tips/TIP-0018/tip-0018.md#nft-output)      |

#### Payload

The  _Transaction Essence_ itself can contain another payload as described in general in  [draft TIP-24](https://github.com/iotaledger/tips/pull/55). The [semantic validity](#semantic-validation) of the encapsulating _Transaction Payload_ does not have any impact on the payload.

The following table lists all the payload types that can be nested inside a _Transaction Essence_ as well as links to the corresponding specification:

| Name        | Type Value | TIP                                                        |
|-------------|------------|------------------------------------------------------------|
| Tagged Data | 5          | [draft TIP-23](https://github.com/iotaledger/tips/pull/54) |

### Unlocks

The `Unlocks` field holds the unlocks unlocking inputs within a _Transaction Essence_.

> The following table lists all the output types that are currently supported as well as links to the corresponding specification. The _Signature Unlock_ as well as the _Reference Unlock_ is specified as part of this TIP.

Reviewer 3: ```suggestion
The following table lists all the output types that are currently supported as well as links to the corresponding specification. The _Signature Unlock_ and the _Reference Unlock_ are specified as part of this TIP.
```

            <td>The Ed25519 signature signing the Blake2b-256 hash of the serialized <i>Transaction Essence</i>.</td>
          </tr>
        </table>
      </details>
    </td>
  </tr>
</table>

##### Unlock syntactic validation

* `Signature` must contain an _Ed25519 Signature_.
* The _Signature Unlock Block_ must be unique, i.e. there must not be any other _Signature Unlock Blocks_ in the `Unlock Blocks` field of the transaction payload with the same signature.

#### Reference Unlock Block

The <i>Reference Unlock Block</i> defines an <i>Unlock Block</i> which references a previous <i>Unlock Block</i> (which must not be another <i>Reference Unlock Block</i>). It **must** be used if multiple inputs can be unlocked via the same <i>Unlock Block</i>. It is serialized as follows:

<table>
  <tr>
    <th>Name</th>
    <th>Type</th>
    <th>Description</th>
  </tr>
  <tr>
    <td>Unlock Type</td>
    <td>uint8</td>
    <td>
      Set to <strong>value 1</strong> to denote a <i>Reference Unlock Block</i>.
    </td>
  </tr>
  <tr>
    <td>Reference</td>
    <td>uint16</td>
    <td>Represents the index of a previous unlock block.</td>
  </tr>
</table>

##### Unlock syntactic validation

> * The _Reference Unlock Block_ at index i must have `Reference` < i and the unlock block at index `Reference` must be a _Signature Unlock Block_.

Reviewer: Only `Reference Unlock Block` is shown here, Alias and Nft Unlock Blocks are missing, of course they're very similar, but they are not mentioned there so it's not clear if the syntactic validation is also required for them/the same

Author: Couple lines above, the Unlock Blocks section starts with the following sentence:

> The following table lists all the output types that are currently supported as well as links to the corresponding specification. The _Signature Unlock Block_ as well as the _Reference Unlock Block_ is specified as part of this TIP.

| Unlock Block Name | Type Value | TIP                                                                                                                   |
| ----------------- | ---------- | --------------------------------------------------------------------------------------------------------------------- |
| Signature         | 0          | [TIP-20](#signature-unlock-block)                                                                                     |
| Reference         | 1          | [TIP-20](#reference-unlock-block)                                                                                     |
| Alias             | 2          | [draft TIP-18](https://github.com/lzpap/protocol-rfcs/blob/master/tips/TIP-0018/tip-0018.md#alias-locking--unlocking) |
| NFT               | 3          | [draft TIP-18](https://github.com/lzpap/protocol-rfcs/blob/master/tips/TIP-0018/tip-0018.md#nft-locking--unlocking)   |

Ofc, there is an error in the sentence, as it should read: _...lists all the **unlock blocks** that are currently supported..._
Anyway, this sentence clearly says that only these two unlock blocks are discussed here, the others are linked. Isn't that sufficient?

It is also mentioned at the sytectic validation rules that:
>  * For each unlock block the following must be true:
>    * Each `Unlock Block Type` must match one of the values described under [Unlock Blocks](#unlock-blocks).
>    * The unlock block itself must pass syntactic validation.

Reviewer: Well yes, I was only looking for `Alias Unlock Block Syntactic Validation` and didn't found it, because it's only linked, so my bad :carlos:

Example:
Consider a <i>Transaction Essence</i> containing the <i>UTXO Inputs</i> 0, 1 and 2, where 0 and 2 are both spending outputs belonging to the same Ed25519 address A and 1 is spending from a different address B. This results in the following structure of the <i>Unlocks</i> part:
| Index | Unlock                                                                             |
| ----- | ---------------------------------------------------------------------------------------- |
| 0     | A _Signature Unlock_ holding the Ed25519 signature for address A.                  |
| 1     | A _Signature Unlock_ holding the Ed25519 signature for address B.                  |
| 2     | A _Reference Unlock_ which references 0, as both require the same signature for A. |

## Validation

> A <i>Transaction Payload</i> has different validation stages, since some validation steps can only be executed when certain information has (or has not) been received. We therefore distinguish between syntactic and semantic validation

Reviewer 3: ```suggestion
A <i>Transaction Payload</i> has different validation stages, since some validation steps can only be executed when certain information has (or has not) been received. We therefore distinguish between syntactic and semantic validation.
```

Author: Added in https://github.com/iotaledger/tips/pull/40/commits/d9370b586a61719331d7fd1fccea365ee5c5d137

The different output types and optional output features introduced by [draft TIP-18](https://github.com/iotaledger/tips/pull/38) add additional constraints to the transaction validation rules, but since these are specific to the given outputs and features, they are discussed for each [output type](https://github.com/lzpap/protocol-rfcs/blob/master/tips/TIP-0018/tip-0018.md#output-design) and [feature type](https://github.com/lzpap/protocol-rfcs/blob/master/tips/TIP-0018/tip-0018.md#output-features) separately.

### Syntactic validation

> Syntactic validation is checked as soon as the transaction data has been received in its entirety. It validates the structure but not the signatures of the transaction. If the transaction does not pass this stage, it must not be broadcast further and can be discarded right away.

Reviewer 3: ```suggestion
Syntactic validation is checked as soon as the transaction data has been received in its entirety. It validates the structure but not the signatures of the transaction. If the transaction does not pass this stage, it must not be broadcasted further and can be discarded right away.
```

> Syntactic validation is checked as soon as the transaction data has been received in its entirety. It validates the structure but not the signatures of the transaction. If the transaction does not pass this stage, it must not be broadcast further and can be discarded right away.

Author: Added https://github.com/iotaledger/tips/pull/40/commits/d9370b586a61719331d7fd1fccea365ee5c5d137

The following criteria defines whether a payload passes the syntactical validation:

* Essence:
  * `Transaction Type` value must denote a _Transaction Essence_.
  * `Network ID` must match the value of the current network.
  * Inputs:
    * `Inputs Count` must be 0 < x ≤ `Max Inputs Count`.
    * For each input the following must be true:
      * `Input Type` must denote a _UTXO Input_.
      * `Transaction Output Index` must be 0 ≤ x < `Max Outputs Count`.
    * Each pair of `Transaction ID` and `Transaction Output Index` must be unique in the list of inputs.
  * Outputs:
    * `Outputs Count` must be 0 < x ≤ `Max Outputs Count`.
    * For each input the following must be true:
      * `Output Type` must match one of the values described under [Outputs](#outputs).
      * The output itself must pass syntactic validation.

>     * The sum of all `Amount` fields must not exceed `Max IOTA Supply`.

Reviewer 3: Maybe also a sentence about Native Tokens count ?

  * Payload (if present):
    * `Payload Type` must match one of the values described under [Payload](#payload).

>     * `Data fields` must be correctly parsable in the context of the `Payload Type`.

Reviewer 3: I don't understand this sentence

Author: Clarified in https://github.com/iotaledger/tips/pull/40/commits/d9370b586a61719331d7fd1fccea365ee5c5d137

### Semantic validation

The Semantic validation of a _Transaction Payload_ is performed when its encapsulating message is confirmed by a milestone. The semantic validity of transactions depends on the order in which they are processed. Thus, it is necessary that all the nodes in the network perform the checks in the same order, no matter the order in which the transactions are received. This is assured by using the White-Flag ordering as described in  [TIP-2](../TIP-0002/tip-0002.md#deterministically-ordering-the-tangle).

Processing transactions according to the White-Flag ordering enables users to spend UTXOs which are created in the same milestone confirmation cone, as long as the spending transaction comes after the funding transaction in the aforementioned White-Flag order. In this case, it is recommended that users include the _Message ID_ of the funding transaction as a parent of the message containing the spending transaction.

The following criteria defines whether a payload passes the semantic validation:
* Each input must reference a valid UTXO, i.e. the output referenced by the input's `Transaction ID` and `Transaction Output Index` is known (booked) and unspent.

> * `Inputs Commitment` must be equal to the BLAKE2b-256 hash of the serialized outputs that are referenced as inputs. The order of the serialized outputs must be the same as their order in `Inputs`.

Reviewer 4: ```suggestion
* `Inputs Commitment` must be equal to the BLAKE2b-256 hash of the BLAKE2b-256 hashes of the serialized outputs that are referenced as inputs. The order of the serialized outputs must be the same as their order in `Inputs`.
```

---
tip: 20
title: Transaction Payload with TIP-18 Output Types
description: Add output types, unlocks, and output features from TIP-18 into Transaction Payload
author: Levente Pap (@lzpap) <levente.pap@iota.org>
discussions-to: https://github.com/iotaledger/tips/pull/40
status: Draft
type: Standards
layer: Core
created: 2021-11-18

> requires: TIP-7 and TIP-18

Reviewer 4: The "replaces" column is missing in this tip. Shouldn't it replace TIP-7 ?

---

# Summary

This TIP proposes a UTXO-based transaction structure consisting of all the inputs and outputs of a transfer. Specifically, this TIP defines a transaction payload for _blocks_ described in [TIP-24](../TIP-0024/tip-0024.md).

# Motivation

> There are several options on how transfers of IOTA coins can be embedded into the actual vertices of the Tangle. In the legacy IOTA protocol, each vertex corresponds to a single transaction, i.e. an input or an output. In order to support atomic transfers, input/output transactions have to be grouped into a so-called bundle and then either the entire bundle is applied to the ledger or none of its transactions.

Reviewer 4: Do we still have the same motivation like in TIP-7?
Not sure if we need to talk about the old legacy IOTA here, if this was already done in the TIP that got replaced by this one.
Might be a bit confusing to the reader that doesn't know the old tech IMO.

                  Describes a deposit to a single address. The output might contain optional features and native tokens.
                </blockquote>
              </details>
              <details>
                <summary>Alias Output</summary>
                <blockquote>
                  Describes an alias account in the ledger.
                </blockquote>
              </details>
              <details>
                <summary>Foundry Output</summary>
                <blockquote>
                  Describes a foundry that controls supply of native tokens.
                </blockquote>
              </details>
              <details>
                <summary>NFT Output</summary>
                <blockquote>
                  Describes a unique, non-fungible token deposit to a single address.
                </blockquote>
              </details>
            </td>
          </tr>
          <tr>
            <td>Payload Length</td>
            <td>uint32</td>
            <td>The length in bytes of the optional payload.</td>
          </tr>
          <tr>
            <td valign="top">Payload <code>optOneOf</code></td>
            <td colspan="2">
              <details>

>                 <summary>Generic Payload</summary>

Reviewer 4: Is the payload really generic?
Even though that we are defining the allowed payloads in the following paragraph, it differs from the way how the Outputs are defined above for example. Could be misleading IMO.

                <blockquote>
                  An outline of a generic payload.
                </blockquote>
                <table>
                  <tr>
                    <th>Name</th>
                    <th>Type</th>
                    <th>Description</th>
                  </tr>
                  <tr>
                    <td>Payload Type</td>
                    <td>uint32</td>
                    <td>
                      The type of the payload. It will instruct the node how to parse the fields that follow.
                    </td>
                  </tr>
                  <tr>
                    <td>Data Fields</td>
                    <td>ANY</td>
                    <td>A sequence of fields, where the structure depends on <code>Payload Type</code>.</td>
                  </tr>
                </table>
              </details>
          <tr>
        </table>
      </details>
    </td>
  </tr>
  <tr>
    <td>Unlocks Count</td>
    <td>uint16</td>
    <td>The number of unlock entries. It must match the field <code>Inputs Count</code>.</td>
  </tr>
  <tr>
    <td valign="top">Unlocks <code>anyOf</code></td>
    <td colspan="2">
      <details>

>         <summary>Signature Unlock</summary>

Reviewer 4: The table looks broken in my browser.

Author: Is it fixed now?

* Additionally, local snapshots can no longer be represented by a list of addresses and their balances, since the ledger is now made up of the UTXOs on which the actual funds reside. Therefore, local snapshot file schemes have to be adjusted to incorporate the transaction hashes, output indices, and then the destination addresses including the balances.

# Rationale and alternatives

* Introducing this new transaction structure allows for extensions in the future, to accommodate new requirements. With the support for Ed25519 addresses/signatures, transaction size is drastically reduced and allows for safe re-signing in case of address reuse. Due to the switch to a complete binary transaction, the transaction size is reduced even further, saving network bandwidth and processing time.
* Other transaction structures have been considered but they would have misused existing transaction fields to accommodate for new features, instead of putting them into a proper descriptive structure. Additionally, those ideas would not have been safe against replay attacks, which deems reusing the old transaction structure, for example for Ed25519 addresses/signatures, as infeasible.

> * Not switching to the new transaction structure described in this TIP would have led to more people losing funds because of W-OTS address reuse and it would prevent extending the IOTA protocol further down the line.

Reviewer 4: Maybe we can also get rid of some rationales because it was already part of TIP-7

---
tip: 20
title: Transaction Payload with TIP-18 Output Types

> description: Add output types, unlocks and output featured from TIP-18 into Transaction Payload

Reviewer 5: ```suggestion
description: Add output types, unlocks, and output features from TIP-18 into Transaction Payload
```
